ogr/ogrsf_frmts/xodr/ogrxodrdatasource.cpp

```diff
@@ -0,0 +1,186 @@
+/******************************************************************************
+ *
+ * Project:  OpenGIS Simple Features for OpenDRIVE
+ * Purpose:  Implementation of OGRXODRDataSource.
+ * Author:   Michael Scholz, German Aerospace Center (DLR)
+ *           Gülsen Bardak, German Aerospace Center (DLR)        
+ *
+ ******************************************************************************
+ * Copyright 2024 German Aerospace Center (DLR), Institute of Transportation Systems
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include "ogr_xodr.h"
+using namespace odr;
+using namespace pugi;
+using namespace std;
+
+CPL_CVSID("$Id$")
+
+OGRXODRDataSource::OGRXODRDataSource()
+{
+    layers = nullptr;
+    nLayers = 0;
+}
+
+OGRXODRDataSource::~OGRXODRDataSource()
+{
+    for (int i = 0; i < nLayers; i++)
+    {
+        delete layers[i];
+        layers[i] = nullptr;
+    }
+    CPLFree(layers);
+}
+
+int OGRXODRDataSource::Open(const char *fileName, char **openOptions,
+                            int bUpdate)
+{
+    VSILFILE *file = nullptr;
+
+    bool updatable = CPL_TO_BOOL(bUpdate);
+    if (updatable)
+    {
+        CPLError(CE_Failure, CPLE_OpenFailed,
+                 "Write or update access not supported by XODR driver.");
+        return CE_Failure;
+    }
+    else
+    {
+        file = VSIFOpenExL(fileName, "r", true);
+    }
+
+    if (file == nullptr)
```

Comment:
Not necessarily. The user might be a GDAL /vsi file.  But I suspect your OpenDrive code only supports regular files. So you should rather use the standard VSIFOpen() or fopen().
Ah I believe I now understand your doc comment about UTF-8 and files. What does  odr::OpenDriveMap xodr(fileName, false); expect?  Because GDAL will provide a UTF-8 filename.